### ISSUE TITLE
org-ref has recently gotten very slow

### ISSUE THREAD

**Issue by notuntoward** [jkitchin/org-ref#468]
I have an 11K line org file that, up until a few weeks ago, opened quickly.  But recently, opening it takes about a minute.  The file has a lot of links to my bibtex file, mostly the standard org bibtex links.  But there are also a few hundred org-ref-cite links, and when I  profile the opening of this file, I see that the cite links are consuming 84% CPU (trace pasted below).  

Is this expected?  Has some new compute-intensive functionality been recently added?

I'm running org-ref 1.1.1 (latest from elpa) on emacs 25.1.1, compiled for windows.

[fileopenprofile.txt](https://github.com/jkitchin/org-ref/files/1106640/fileopenprofile.txt)

**Comment by Titan-C** [comment]
I have a 5K org file. I don't experience yet a slow down on opening. Nevertheless while typing a reference say `ref:my-fig` emacs becomes quite unresponsive. Like the font locking routine that displays it in a bold pink font goes crazy finding the reference. Once I manage to fully type the reference it is then displayed in an underlined red font and I can continue to type at normal speed.

**Comment by jkitchin** [comment]
That is probably from org-ref-show-broken-links. Try (setq org-ref-show-broken-links nil) and see if it fixes this. For small files I don't see any significant issue, but for large files I can see it causing this problem. You can permanently disable it by putting the setq above in an init file, or as a file local variable (I haven't tested that, but it should be ok).

This variable enables a conditional face on org-ref links so they show in red if they are not valid, e.g. can't find the cite key, multiple labels, or a ref that does not point to a label.

To avoid the typing issue you refer to @Titan-C you can type C-u C-c ] to get a list of labels to complete on.

**Comment by notuntoward** [comment]
Thanks, yes, disabling show-broken-links has vastly sped up org file opening.  org-ref's CPU load is now pretty insignificant.  

The broken links display really helped me pick some lint out of my org file, though.  Could the broken link check be called within some kind org file integrity checker?

**Comment by jkitchin** [comment]
You can see the broken links by running M-x org-ref. It is the closest to a lint check for org-ref as it checks all the links, and several other things in the document that might cause problems.

I have been trying to think of a way to speed up the broken link checking, but so far all my ideas rely on caching and text properties. It will take some creative solutions to fix it in large documents.

**Comment by abradd** [comment]
I was also experiencing long delays when opening large files. It felt like the response time dropped dramatically at one point relative to the number of new links I was adding.

Another issue I found was that typing on lines with org-ref refs or labels was slow.

Setting org-ref-show-broken-links to nil helped a lot in both cases. 

I have attached my profile when opening a large file.
[org-ref-profile.txt](https://github.com/jkitchin/org-ref/files/1133197/org-ref-profile.txt)

I have also attached the profiler report for typing on a line with an org-ref ref.
[org-ref-profile-typing.txt](https://github.com/jkitchin/org-ref/files/1133198/org-ref-profile-typing.txt)

fontify doesn't seem to consume as much CPU as the original post, but it is enough to produce noticeable lag. The garbage collector seems to be doing a lot of work that can be reduced by setting org-ref-show-broken-links to nil.

Profiles were generated on Emacs 25.1.1 (though I have tested on 25.2.1) with org v9.0.9 and org-ref v1.1.1 on macos 10.12.5.

**Comment by notuntoward** [comment]
I've noticed a couple more places where org-ref has gotten slow:

1. **Emacs startup**.  I wondered why my emacs startup had gotten so slow.  The emacs `esup `packages shows that half of the startup time (5.7 secs) is consumed by org-ref.  Here's the esup output:

```
org-ref.elc:15  5.710sec   51%
(byte-code "\301\302!\210\301!\210\303\304!\207" [org-ref-completion-library require org-ref-core provide org-ref] 2)
```
2. **First `C-c l` in a .bib file**.  The first time I type `C-c l` on a bibtex entry, it takes about 6 seconds before I see the org-ref question in the minibuffer.  I mean the one that asks "which function for creating org-link?"  The bibtex file has about 46K lines.

**Versions**
emacs: 25.2.1, windows 64 bit
org-ref: 1.1.1
org: 9.1.2

**Comment by jkitchin** [comment]
How did you run the esup command? When I run it org-ref takes only 1 second to load. pdf-tools took about 5 seconds to load. and the bootstrap.el takes a while too, probably because it fetches updates from MELPA.

I can't reproduce #2 in a bib file that has about 30K lines. It is very fast for me.

**Comment by notuntoward** [comment]
I installed the esup package and then ran M-x esup. 

This must be something specific to my org setup.  Do you have any advice for how I could debug further?

**Comment by jkitchin** [comment]
I guess the slow part is loading your cache file. For me that seems to be about 17% of the load time for org-ref, and my bibfile only has ~30K lines. I am not sure the best way to reduce that. One is to use another backend that doesn't use this cache (although I made the cache because parsing the bibtex file every time is slow). Once the cache is in memory, it is fast until you have changed the bibtex file and it has to be updated. Breaking the bib file into smaller files would help  on that, as only the changed files would be updated.

Some other alternatives might include using an idle timer to load it, or an async process.

What I did to figure this out was to open emacs with:
emacs -q /path/to/scimax/init.el

Then I manually executed the lines up to the use-package for org-ref. Before that I ran M-x profiler start, and then ran the usepackage line, and then ran M-x profiler-report.

**Comment by notuntoward** [comment]
Hmm, I just tried your speed measuring approach and see the same thing you do.  Maybe the esup package is marked obsolete for a reason.

I'll look into this some more...

**Comment by notuntoward** [comment]
After reorganizing my `.emacs`  around `use-package`, `org-ref `is dramatically faster.  I tried several things but don't know which combination is responsible for the improvement.  Anyway, `org-ref `speed is no longer a problem for me so I'll close this.

**Comment by junwei-wang** [comment]
What is the reasonable time for loading a [bib](https://github.com/cryptobib/export/blob/master/crypto.bib) of 25 MB.

**Comment by jkitchin** [comment]
My emacs opens this file almost instaneously. 

Note that the first time you use it with org-ref, it takes maybe 30 seconds to create a cache file. After that though, it is well under a second to insert citations. The cache is rebuilt each time you modify crypto.bib.

**Comment by junwei-wang** [comment]
@jkitchin Thank you for your reply. 

It takes me about 2 minutes to load this bib with org-ref. And this loading is triggerred each time I invoke `org-ref` command.

**Comment by jkitchin** [comment]
What backend are you using? If it is helm-bibtex, that may be the issue. One of the reasons I wrote the ivy backends is that I found helm-bibtex too slow as it parses the file each time. The ivy backends are much faster.

**Comment by junwei-wang** [comment]
I have `helm-bibtex` `bibtex-completion` `org-ref` `org-roam-bibtex` installed. 

For now, I am not quite clear in their relationships and I don't really know how to change backend of `org-ref`.

**Comment by junwei-wang** [comment]
If you don't mind, could you please share your config?

**Comment by jkitchin** [comment]
I think this is what I use:

```
(setq org-ref-completion-library 'org-ref-ivy-cite)
(require 'org-ref)
```

**Comment by mcraveiro** [comment]
For anyone experiencing this issue now (May 2021): I too was experiencing dramatic slow downs with org-ref, though perhaps my case may be a bit extreme - I have a single org-file with > 500Kb and a bib file of 56K. I tried profiling and I could not really find anything obvious in the traces. However, when I did the suggested ```setq```, my performance improved immediately:

```lisp
(setq org-ref-show-broken-links nil)
```
Many thanks, this was extremely helpful. I don't really use broken links that much, so I'll just keep them disabled. It would be nice if these could show up when running ```org-lint```, which I do on demand. For now I use @jkitchin's suggestion and run ```org-ref``` manually.

Cheers
